### PR TITLE
[onedpl] Merge Path algorithm for oneDPL merge algorithm (SYCL backend)

### PR DESCRIPTION
**Pull request by MikeDvorskiy**
[onedpl] Merge Path algorithm for oneDPL merge algorithm (SYCL backend).

This is a new implementation of the oneDPL merge algorithm  (merging two sorted ranges) based on new approach known as Merge Path parallel merging. It gives a significant speed up comparing to the current oneDPL implementation.
The main idea  - each work-item is searching for the start points in range 1 and range 2. After each work-item does M steps of sequential merging, where M = (N1+N2) / p, where  p -  number of threads. So, we have got well balanced parallel merging.

### CODE REVIEW

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1489]: spelling

```suggestion
        auto __q = __i_elem; //diagonal index
        auto __n_diag = __q < __n_2 ? __q : __n_2; //diagonal size
```

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1494]: `__val1` is a bool as it is the result of the `<` comparison, then we compare it to the diagonal index `__v2`?

Is there a better name for `__val1` to indicate what it is supposed to represent?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1505]: Can this be `auto __n_diag = std::min(__n_2 - q, __n_1)` for better clarity?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1489]: Can this be `auto __n_diag = std::min(__q, __n_2)`  for better clarity?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1531]: `__i` needs to be explicitly typed `__Index2` to accommodate `__n`.    This appears in a few more places below.


`_Size` should be included in the numbered `_SizeX`, types.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (mmichel11) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1494]: Should this comparison use the provided binary comparator?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (mmichel11) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1482]: Can `Rng1` and `Rng2` be const references here? We also may want to rename to `_Rng1` and `_Rng2` for consistency.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (mmichel11) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1604]: We can probably use `__dpl_ceiling_div` in this line.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (mmichel11) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1602]: Should we use a smaller data type?

**Review by danhoeflinger — COMMENTED** [review]
Still understanding the algorithm, but adding these now to prevent redundant comments, as others are reviewing.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (mmichel11) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1611]: Could there be any benefit to loading a work group's data into SLM with a coalesced access pattern before passing it to this function? Since work items can't coalesce their accesses in `__find_start_point`, putting it into SLM first may be beneficial.

Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1611]: If I'm understanding the algorithm correctly, this only works if ALL input data fits into SLM.  

Also, we don't have any guarantees on the data here (and how they will be merged), its possible that you may be loading all the input data to end up using only a part of it within the work group.  It would take some investigation to understand if this is of benefit for a "normal" use case, and you would need to come up with a strategy to handle the case where only part of the input sequence fits into SLM, and/or a backup where you don't use SLM.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1549]: Seems like these could be `const auto&`  to save some copies.

**Review by danhoeflinger — COMMENTED** [review]
Some more minor things to fix.

At a higher level, I'm still not quite understanding the `__find_start_point` algorithm yet, but I'm getting there.  I'm confused by the `<` comparison between the `bool` (`__val1`) and the diagonal index (`__v2`).  I'll keep looking at this tomorrow.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1494]: No, should not. Here we compare juts zeros and ones.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1611]: Probably, is to try to use SLM makes sense... But in that case we have got the additional n1+n2 reading and n1+n2 additional writing. Together with fact that is almost no computation - just fast binary search + m steps of merging, it is no obvious that we have got a performance gain....
BTW, I tried pre-caching  'chunk' elements form the first seq and 'chunk' elements form the second seq to private memory (registers), and got the performance slow down to 10%.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1521]: what do u mean?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1494]: '__v2' is not diagonal index - this is a value to be found. We are searching for '1'.

Comment (dmitriy-sobolev) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1494]: As far as I see, it is possible to get integer an integer underflow when `__q` and `v1` are `0`.

Comment (dmitriy-sobolev) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1494]: I agree with Matthew. Moreover the element type would have to support `<` operator. There is no such a requirement for C++ merge. Please note, that `merge.pass.cpp` do not use predicates other than `std::less`.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1521]:
```suggestion
template<typename _Rng1, typename _Rng2, typename _Rng3, typename _Index1, typename _Index2, typename _Index3,
         typename _Size1, typename _Size2, typename _Size3, typename _Compare>
```

For better consistency of naming.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: I believe they are talking about 

`auto __val1 = __rng1[__q - __v1 - 1] < __rng2[__v1];`
->
`auto __val1 = comp(__rng1[__q - __v1 - 1], __rng2[__v1]);`

Not the comparison on the next line.  Does this not compare values of the input ranges?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1524]: Yes, I can see your point, It would be bad to change the indices of the related template arguments to not match, or mislead that the chunk size as corresponding to the `_Rng3`.  

It may be more clear that it is intentional if you reorder the parameters to have `_Size __chunk` second to last.  However, I'm fine with however you decide to do this.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: I see.... here- yes, it compares the values of of the input ranges, right, thanks.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1533]: I suggest trying it after the upcoming release. I can make a "reference in new issue"...

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: thanks, that helps.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1524]: "_Size __chunk second to last." also breakers a call semantic... 
Indeed,
```
__serial_merge(const _Rng1& __rng1, const _Rng2& __rng2, _Rng3& __rng3, _Index1 __start1, _Index2 __start2,
               _Index3 __start3, _Size __chunk, _Size1 __n1, _Size2 __n2, _Compare __comp)
```
Tells:
"Does serial merge the data from rng1(starting from start1) and rng2(starting from start2) and writting to rng3 (starting from start3) in '__chunk' steps , but doesn't exceed of total size of  sequences  - n1 and n2."

May be it is better if '__chunk' will be at the first palce:
"Does  '__chunk' steps of serial merging of  the data from rng1 and rng2....."

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Actually, I'm still not convinced about `__v2`.  Is it not just the next element of the `counting_iterator`?

It seems like you are wanting the input iterator sequence to be already transformed into  `__rng1[__q - idx - 1] < __rng2[idx]` where `idx` is  is the `counting_iterator` element, but that isn't the case as written.  

I think you would need to zip together the input ranges, have a transform view and a functor which applies the above expression, then pass in that transform iterator into the `lower_bound` algorithm to get the result you are looking for here.

Unless I'm missing something, `__v2` is just the next element of the input counting iterator after `__v1` and it has not been similarly transformed.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: You've changed the type to `::std::uint8_t`, which I believe will work if `__serial_merge` is called in the context of this algorithm.  However, if this is the intention, then we should probably just remove the `_Size` template argument, and just explicitly type `__chunk` as `::std::uint8_t`.  This solves the other discussion as well.

If instead we want `__serial_merge` to be general and available for use in other algorithms, then `__i` needs to be one of `_Size`, `_Index2` or `_Size2` to make sure there is no overflow.  Ideally it could be typed as the smallest of those types, and we could write some compile time helpers to accomplish that.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Cppreference tells about std::lower_bound:
```
Returns an iterator pointing to the first element in the range [first, last) such that element < value (or comp(element, value)) is false, (i.e. that is greater than or equal to value), or last if no such element is found.

The range [first, last) must be partitioned with respect to the expression element < value (or comp(element, value)), i.e., all elements for which the expression is true must precede all elements for which the expression is false. A fully-sorted range meets this criterion.
```

Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Sorry, I've been thinking about it wrong.  My mistake.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (mmichel11) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: I think to have any performance benefit with SLM, we will have to look at the GPU Merge Path variant of Merge Path which uses a tiling strategy. This can probably be a future optimization if it is difficult to implement since this implementation already provides significant performance benefit.

**Review by dmitriy-sobolev — APPROVED** [review]
LGTM. Please, double check the CI results before merging. You also might want to get an approval from the other reviewers.

**Review by dmitriy-sobolev — APPROVED** [review]
Reapproving after some minor changes (e.g. `__n_1` is renamed to `__n1` for consistency).

**Review by danhoeflinger — DISMISSED** [review]
LGTM

**Review by dmitriy-sobolev — APPROVED** [review]
Reapproving after some formatting changes.

**Pull request merged**
